### ISSUE TITLE
GPT2 not use mha kernel?

### ISSUE THREAD

**Issue by lk137095576** [NVIDIA/TensorRT#2736]
I profile the GPT2 in huggingface/demo with nsight computer, but I found that TRT not use mha kernel, ie FasterTransformer/FlashAttention , it is the splited matmul and softmax.
GPU:A100   cuda:11.7  trt: 8.5.3.1
<img width="543" alt="ncu" src="https://user-images.githubusercontent.com/26320143/223069535-7b0d0a3a-aad3-439a-a374-e01bc53a01a0.png">
If the FasterTransformer/FlashAttention  is slower than upper kernel? 
Or  you have not enought time to optimize it?

**Comment by zerollzeng** [comment]
@nvpohanh ^ ^

**Comment by nvpohanh** [comment]
This support will be added in TRT 8.6

**Comment by lk137095576** [comment]
> 

Do you have any plans to release version 8.6？

**Comment by lk137095576** [comment]
@nvpohanh  I try 8.6, but it also not fuse.

**Comment by nvpohanh** [comment]
@lk137095576 Could you use `release/8.6` branch and share your command?
https://github.com/NVIDIA/TensorRT

The command I used: 

```
python3 run.py benchmark GPT2 trt --working-dir temp --variant gpt2 --fp16 --batch-size=<BS> --input-seq-len=<INSEQLEN> --output-seq-len=<OUTSEQLEN> --duration=60 --enable-kv-cache

python3 run.py benchmark GPT2 trt --working-dir temp --variant gpt2-large --fp16 --batch-size=<BS> --input-seq-len=<INSEQLEN> --output-seq-len=<OUTSEQLEN> --duration=60 --enable-kv-cache

python3 run.py benchmark GPT2 trt --working-dir temp --variant gpt2-xl --fp16 --batch-size=<BS> --input-seq-len=<INSEQLEN> --output-seq-len=<OUTSEQLEN> --duration=60 --enable-kv-cache

python3 run.py benchmark GPT2 trt --working-dir temp --variant EleutherAI/gpt-j-6B --fp16 --batch-size=<BS> --input-seq-len=<INSEQLEN> --output-seq-len=<OUTSEQLEN> --duration=60 --enable-kv-cache
```

**Comment by lk137095576** [comment]
@nvpohanh 

I git clone the release/8.6 branch and prof the command as:
` ncu  --section MemoryWorkloadAnalysis  -o run_demo.ncu-rep  python3 run.py benchmark GPT2 trt --working-dir temp --variant gpt2 --fp16 --batch-size=1 --input-seq-len=1024 --output-seq-len=1024 --duration=60 --enable-kv-cach `

but I got the result: It looks like the MHA of three kernel.
![image](https://user-images.githubusercontent.com/26320143/226813980-a35a9246-72bc-4dda-8f79-61453a7d7b67.png)

**Comment by nvpohanh** [comment]
In GPT2, the MHA in the context phase (the very first decoding iteration) is not fused. Only the MHA in the generation phase (the second decoding iteration and after) is fused. Could you run with `--input-seq-len=1 --output-seq-len=1024` and check if the kernels have "mha_xx" kernels?

**Comment by lk137095576** [comment]
I profile with nsys as follows:
`nsys profile --stats=true  python3 run.py benchmark GPT2 trt --working-dir temp --variant gpt2 --fp16 --batch-size=1 --input-seq-len=1 --output-seq-len=1024 --duration=60 --enable-kv-cach > nsys_gpt_demo.log`
but also not get kernel "mha_xxx".
[nsys_gpt_demo.log](https://github.com/NVIDIA/TensorRT/files/11046341/nsys_gpt_demo.log)
I try the ncu, but get the same result.

**Comment by nvpohanh** [comment]
@lk137095576 Could you add `--verbose` flag and share the logs? Also, have you installed TRT 8.6 EA build?

**Comment by lk137095576** [comment]
1.  yes, I installed TRT8.6 EA with "TensorRT-8.6.0.12.Linux.x86_64-gnu.cuda-11.8.tar.gz"
2. command trtexec  as:
`&&&& PASSED TensorRT.trtexec [TensorRT v8600] # trtexec`
3. add "--verbose", get the log:
"(base) root@xxxxxxx:/home/xxxx/TensorRT_86/demo/HuggingFace# nsys profile --stats=true  python3 run.py benchmark GPT2 trt --working-dir temp --variant gpt2 --fp16 --batch-size=1 --input-seq-len=1 --output-seq-len=1024 --duration=60 --enable-kv-cach --verbose > nsys_gpt_demo.log
[2023-03-23 02:46:34,170][OSS][WARNING] input_profile_max_len is not provided, default to 1024
[2023-03-23 02:46:34,170][OSS][WARNING] output_profile_max_len is not provided, default to 1024
[2023-03-23 02:46:37,429][OSS][INFO] Reading and loading engine file temp/GPT2/gpt2/GPT2-gpt2-fp16-kv_cache/decoder/GPT2-gpt2-fp16-kv_cache-bs1-inseq1-outseq1024.engine using trt native runner.
[2023-03-23 02:46:39,388][OSS][DEBUG] Number of profiles detected in engine: 50
[2023-03-23 02:46:52,291][OSS][DEBUG] Measuring inference call with warmup: 3 and number: 1 and iterations 10 and duration 60 secs
[2023-03-23 02:46:52,300][OSS][DEBUG] Warmup times: [0.005615461966954172, 0.0016831679968163371, 0.001658021006733179]
[2023-03-23 02:47:52,309][OSS][DEBUG] Measuring inference call with warmup: 3 and number: 1 and iterations 10 and duration 60 secs
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
[2023-03-23 02:48:02,189][OSS][DEBUG] Warmup times: [3.3311963339801878, 3.270483865053393, 3.2786741609452292]
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
"

**Comment by nvpohanh** [comment]
@zerollzeng Is this something you can try to repro and file an internal tracker for? Thanks

@lk137095576 Which GPU are you using?

**Comment by lk137095576** [comment]
@nvpohanh NVIDIA A100 80GB PCIe

**Comment by nvpohanh** [comment]
We have identified the issue and will fix this in TRT 8.6 GA release.

Meanwhile, could you check T5 and see if the MHA is fused in T5? Thanks

**Comment by lk137095576** [comment]
@nvpohanh  T5 have no MHA fuse too.

**Comment by ttyio** [comment]
closing since now the huggingface demo moved to tensorrt-llm, thanks all!
